perma_web/conftest.py

```diff
@@ -662,7 +702,119 @@ def f(customer):
     return f
 
 
-### fixtures for testing utils ###
+### For working with links ###
+
+@pytest.fixture
+def complete_link_factory(completed_capture_job_factory, primary_capture_factory, screenshot_capture_factory):
+    def f(link_kwargs=None, primary_capture=True, screenshot_capture=True):
+        if link_kwargs:
+            link = completed_capture_job_factory(**{
+                f"link__{k}": v
+                for k, v in link_kwargs.items()
+            }).link
+        else:
+            link = completed_capture_job_factory().link
+
+        if primary_capture:
+            primary_capture_factory(link=link, url=link.submitted_url)
+        if screenshot_capture:
+            screenshot_capture_factory(link=link)
+
+        return link
+    return f
+
+
+@pytest.fixture
+def complete_link(complete_link_factory):
+    return complete_link_factory()
+
+
+@pytest.fixture
+def complete_link_without_capture_job(complete_link):
+    complete_link.capture_job.delete()
+    try:
+        complete_link.capture_job
+    except CaptureJob.DoesNotExist:
+        pass
+    return complete_link
+
+
+@pytest.fixture
+def deleted_link(complete_link_factory):
+    link = complete_link_factory({
+        "cached_can_play_back": False
+    })
+    link.safe_delete()
+    link.save()
+    return link
+
+
+@pytest.fixture
+def deleted_capture_job(deleted_link):
+    # Capture Jobs are marked as 'deleted' if a user deletes that GUID
+    # before the capture job is finished
+    job = deleted_link.capture_job
+    job.status = 'deleted'
+    job.save()
+    return job
+
+
+@pytest.fixture
+def failed_capture_job(in_progress_capture_job):
+    in_progress_capture_job.mark_failed("Something went wrong.")
+    return in_progress_capture_job
+
+
+@pytest.fixture
+def memento_link_set(link_factory):
+    domain="wikipedia.org"
+    url = f"https://{domain}"
+
+    today = timezone.now()
+    a_few_days_ago = today - relativedelta(days=5)
+    within_the_last_year = today - relativedelta(months=6)
+    over_a_year_ago = today - relativedelta(years=1, days=2)
+    three_years_ago = today - relativedelta(years=3)
+
+    links = [
+        link_factory(
+            creation_timestamp=time,
+            submitted_url=url,
+            cached_can_play_back=True,
+        ) for time in [today, a_few_days_ago, within_the_last_year, over_a_year_ago, three_years_ago]
+    ]
+
+    return {
+        "domain": domain,
+        "url": url,
+        "links": [
+            {
+                "guid":link.guid,
+                #
+                "timestamp":  link.creation_timestamp
+            } for link in reversed(links)
+        ]
+    }
+
+
+### For testing email ###
+
+@pytest.fixture
+def email_details():
+    return {
+        "from_email": 'example@example.com',
+        "custom_subject": 'Just some subject here',
+        "message_text": 'Just some message here.',
+        "refering_page": 'http://elsewhere.com',
```

Comment:
I wish the key here were `referring_page`, but that would require changes elsewhere; maybe I'll make a PR.

Reply:
Thanks for these!! 

I can make that change, hold one...

Reply:
Done! I considered intentionally making a typo in the commit message, like `refferring` 😁